### PR TITLE
Add TF id member to DataHeader

### PR DESCRIPTION
**Pull request by mkrzewic**
Since all data parts are equipped with this there is enough reason to move this into the DataHeader, as was envisioned and agreed. Please do comment on the exact layout of this, for now it is long uint. 8 bytes should be enough for everybody.

### CODE REVIEW

**Comment by ktf** [comment]
This will mean we need to change DPL DataDescriptorMatcher to pick up the information from the correct place, correct? Or at least we will need to duplicate it. What about the length (in some arbitrary unit of time)? Can we have that as well, please?

**Comment by mkrzewic** [comment]
The way I undestand it now: the tfid is now needed by something external to the dpl. Therefore, to avoid dependencies, code complication and possibly allocation gymnastics (as noted by @ironMann ) the tfid should go in there. I'm not opposing adding more stuff in the data header, just trying to be a minimalist:)
What would be the benefit/use case of putting the time duration in DataHeader? Especially since it would be somewhat arbitrary? Would it be required for all the data blocks (as in: for routing or some other critical action)?

**Comment by ktf** [comment]
If the only reason is minimalism, what you propose is effectively forcing an extra header to be added at a later stage (with all the overhead it implies) to spare 8 bytes. DPL does need to support some sort of "duration" for messages in order to cache things that spans more than one timeframe (conditions, histograms) or that are less than one timeframe long (e.g. data created on top of triggered detectors).

**Comment by ktf** [comment]
to be clear, it does not have to be filled with the actual durantion (albeit it would make sense, IMHO), it's enough that we have something like `0x0000 0000 0001 0000` to indicate it's a timeframe duration and then DPL can take care to renormalize for the longer than a timeframe / shorter than a timeframe cases.

**Comment by ironMann** [comment]
The TF ID field proposed here holds the absolute value of the TF ID. This value is generated by the trigger and supplied with corresponding raw data. Also, it can be faked by the simulation. It is "calculated" from the orbit number, but can jump. The time cannot be derived directly from this value, without someone maintaining a database.  It's only guarantied to be unique within a run. 
How I see It, all results of processing (outputs) must have the same value of TFID as the corresponding input data. Again, all headers of the input TF will have the same TFID value, as well as all headers of the corresponding CTFs.

**Comment by mkrzewic** [comment]
@ktf would a flag be enough? or a couple of bits maybe? or do you need 64 bits? we still have 31 bits unused in the BaseHeader, we could make some of them "user" definable or something.

For things that span longer than 1 timeframe an additional header (or other type of metadata) does not seem like a bad idea - it probably should contain a precise information about the timeframes it spans or so (tfid would be invalid for those, some sentinel value maybe).
For things that are smaller than the entire timeframe - aren't they just contained in the timeframe and simply part of it? (i.e. they would have the same valid tfid). But also here some additional special metadata should be there (maybe an additional header, maybe in the data itself, like trigger type, timestamp etc.). One could argue that sorting that out is the business of user logic, e.g. the TOF processing has to eat all the parts anyway so the framework/dispatcher/whatever does not need to know about TOF specifics.

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 1]
```diff
@@ -628,6 +628,7 @@ struct DataHeader : public BaseHeader {
   using SplitPayloadIndexType = uint32_t;
   using SplitPayloadPartsType = uint32_t;
   using PayloadSizeType = uint64_t;
+  using GlobalIDType = uint64_t;
 
   //static data for this header type/version
   static constexpr uint32_t sVersion{1};
```
Comment (ironMann) [inline DataFormats/Headers/include/Headers/DataHeader.h:634]: sVersion should probably be bumped.

Comment (mkrzewic) [reply DataFormats/Headers/include/Headers/DataHeader.h]: indeed, in the past we got away without bumping this, it's a good moment to start:)

File: DataFormats/Headers/include/Headers/DataHeader.h [hunk 2]
```diff
@@ -628,9 +628,10 @@ struct DataHeader : public BaseHeader {
   using SplitPayloadIndexType = uint32_t;
   using SplitPayloadPartsType = uint32_t;
   using PayloadSizeType = uint64_t;
+  using GlobalIDType = uint64_t;
```
Comment (shahor02) [inline DataFormats/Headers/include/Headers/DataHeader.h:631]: @mkrzewic  actually, uint32_t would be enough since the orbit is anyway limited by 32 bits.

Comment (mkrzewic) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: I think it would be nice if it could be unique globally, for e.g. processing entire data sets. maybe it can also contain some starting point (run number, run start timestamp or something). We have been talking about it forever but no agreement yet...

Comment (shahor02) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: I don't think anybody will maintain a global TF ID trough data dating, moreover that the TF length might not be fixed. More universal identification would be provided by the run_number (which allows to access orbit/bc at the start of run) + current orbit/bc, but I am not sure what is the role of this info in the transient data header.  
The TF ID with ID unique within the run would be useful for synchronization, but what would you do with extra info?

Comment (ironMann) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: I agree, the name "GlobalIDType" seems misleading. TimeFrameID is more clear and to the point. AFAIK, it's only unique within the data set of a single run. Anyone requiring absolute time of the data must look somewhere else for it. Aim of having TFID in the header is to match all data peaces during the dataflow and processing and the final CTF merger.

Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: Well. Having the time is a requirement from Data Processing, so you will have to provide it in any case, given it's STFB which interfaces with the trigger. Any reason for not using the time as timeframe id? As I understand it, it is provided by the readout.

Comment (ironMann) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: What STFB knows is the following: the TFID, and the first orbit of the TF. 
CTP to my knowledge does not distribute time. How to work out the absolute "time" from this has not been discussed in this context. Should the processing requires more information this needs to be discussed with all relevant subsystems.

Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: So can we use the first orbit of the TF for the ID, which is time-like beside an offset (correct?)? What data processing needs is a monotonically increasing quantity where equal distances in the value correspond to equal distances in time.

Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: We can then discuss who is responsible to deliver the offset for the run and the scale in a separate meeting. I would still find it useful to have it in the header so that we do not need two extra parameter to provided by control / some database to query CCDB.

Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: Duration could also be (last orbit - first orbit) to start with.

Comment (shahor02) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: @ktf then what is wrong with simple TF counter? Or think about the CCDB access on the framework level?

Comment (shahor02) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: ok, got @ktf messages after asking about the CCDB. But you have the time in the DataProcessingHeader, can't you use that

Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: yes, i can use that, but:
1) why not removing the need for a separate header?
2) who will fill the DataProcessingHeader?

Comment (shahor02) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: if you want to merge headers, it is fine, but given that the timestamp for the CCDB access will be needed also during async. processing, the absolute time stamp should be extractable  from the TF data themselves and not expected to be provided by additional source via transient headers.
In principle, the info of the orbit/bc at start_of_run + absolute its time combined with the 1st orbit of the TF is enough to build the abs. timestamp of the TF. The run start info is supposed to come from the CTP (and must be stored in the CCDB). At the moment I have in the GRP object only the abs. time of (simulated) run start, can add also the orbit/bc.

**Comment by ktf** [comment]
To be pragmatic and go forward, I would merge this so that we can have at least a unique counter which is understood by both DPL & STFx. I assume this is at least vaguely monotonic, right?

**Comment by shahor02** [comment]
I would propose to change it explicitly to 
````
uint32_t orbit; // orbit in the run
uint32_t nObits; // lenght it orbits, in case you need duration
uint64_t time; // absolute time in \mus
````
The absolute time would be needed only for ccdb query w/o checking the payload.

**Comment by ironMann** [comment]
It does not make much sense to add all of this to the DataHeader. Not every peace of data has the notion of orbit. Further, raw data of a TF will be split into multiple o2 messages. Should each of them have the precise orbit and time of the actual HBF? ... or the orbit and time is to be same for all messages and it's taken from the very first HBFrame in a TF? If so, why duplicate this in every single message. 
Can't we have a separate timing metadata payload in every TF, but just once. surely that would suffice.

File: DataFormats/Headers/include/Headers/DataHeader.h (hunk 2 above)
Comment (ktf) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: @shahor02 I agree for the async phase, it's easy to extract a time / duration from the file the time and use it in DPL. The problem for the sync phase is that I do not have any way to extract such information if the STFB does not provide it to me in a readily available manner. Clearly we do not want to scan the whole timeframe payload, nor we want to query a DB just to extract a time to start the processing.

Comment (ironMann) [reply DataFormats/Headers/include/Headers/DataHeader.h:631]: I agree with what @shahor02 proposes. However, @kft, I cannot provide the absolute timing information. Component that issues start of the run (possibly restarting the global orbit counter) must make this information available. Iguess this it CTP or DCS. Until STFBs see any data it could be milliseconds, and the absolute time could differ across all FLPs.

**Comment by shahor02** [comment]
@ironMann this would amount to have a special TF header, with the content I had above. I anyway wanted to introduce it for the CTF, we can also attach it to the DPL header stack. Up to @ktf to tell how much having a special TF header will complicate the DPL processing.

**Comment by ktf** [comment]
All I am asking is that all the bits which might be processed separately in the DAG of DPL (i.e. any O2Message, which is the basic unit of our computation) to have a unique identifier of the timeframe they belong to which is time based, in some unit of time (e.g. first orbit of the timeframe in question), and not time agnostic.

This would still be `UniquelyIdentifiable` which to my understanding is the only requirement that you have for DataProcessing, so it would be fine for you, and it would allow me to drop one header and simply rely on the DataHeader. Basically what I am saying is that in order to save 16 bytes per O2Message (in the worst case) you are forcing me to copy and add a full extra header which will always have to be there (since processing will also happen of the FLP for at least some sub-detectors).

Why is "some unique ID in 64 bit" better than "the start time of the timeframe (in your preferred unit)" to identify a timeframe?

**Comment by shahor02** [comment]
@ktf but apart from UniquelyIdentifiable (within the run) you want also to have absolute time for the ccdb query w/o analyzing the payload, right? For the former just the orbit would suffice, not sure why @ironMann says that not every piece of data has orbit: I think all detectors have an RDH at least in the beginning of TF, so the 1st HB orbit can be picked up by the Stfbuilder and then propagated further in the header (even if the RDH is dropped before the Stfsender).
The abs. time is more tricky, essentially the Stfbuilder will need to know at least the abs.time of the start of run (orbit), I don't know how this can be organized.

**Comment by ironMann** [comment]
My point is only this: the time frame, as a collection of diverse data, needs an ID. From the previous, this can be the value of the first HBF orbit. It fits the bill. Each end every single O2 message will have the same TFID. Calling this value `uint32_t orbit` as proposed above is misleading as there will exist O2 messages, e.g. for HBF(tfstart+x) that where the header will say orbit=start, and the data will actually have the value RDH::orbit=start+x. For that reason I propose to call this field TF_ID (or at least first_Orbit)
@kft, would having such value in the header of every single O2 message help with the graph processing?

What you actually need is the absolute start time of TFs, if i understand the ccdb functional requirement correctly. For this you need a central place to query: give me the absolute time for this orbit. IMO, that data can only come from CTP, the place which actually started the run, reseted the orbit counter. Having both, the orbit and the abstime is just redundant.

What would actually be convenient is for the orbits to count since a fixed time epoch (and never to reset or wrap; would need some more bits than current 32). Then anyone would be able to work out the absolute time without extra information. I'm curios if this was ever discussed...

**Comment by shahor02** [comment]
@ironMann ok, we can call it tfOrbit. 
The absolute time is in additional requirement, needed for the ccdb access, though indeed, it alone would suffice also for the TF identification.
But using global orbit since epoch will not work since the orbit duration is not constant: in a long run the time and the orbit counter would diverge.

**Comment by ktf** [comment]
If TFID is really tfOrbit, that's fine with me. That's what I was proposing the whole time. 
Sorry if I was not clear. If you rename that we can merge this immediately and separate the rest of the discussion below.

In principle one would need also orbitInTF := (RDH::orbit - tfOrbit) if DPL needs to handle below timeframe granularity (e.g. for triggered detectors).

If we cannot use:

```
tfOrbit*gTFPeriod + gTimeBeginOfTheRun
```

to determine the absolute time, as @shahor02 seems to imply above, then either we use an approximation to query CCDB or one would need some sort of tfOrbitTimestamp.

To summarize:

```C++
int32_t tfOrbit;     // as described above. Needed to process the DPL graph with timeframe granularity.
int16_t orbitInTF; // as described above. 0 if RDH::orbit not available. Needed to process the DPL graph with subtimeframe granularity, if needed.
int16_t duration;  // duration of the period associated to the O2Message, in orbits.  NUM_ORBITS_TF if not set. Needed to process the DPL graph with subtimeframe granularity, if needed.
int64_t tfOrbitTimestamp; // absolute time, if it cannot be easily derived from the orbit or if the approximation is not good enough.
```

**Comment by shahor02** [comment]
@ktf 

> If we cannot use: ``tfOrbit*gOrbitPeriod + gTimeFirstOrbitOfTheRun``
What I say is the opposite: the gOrbitPeriod is not really a constant on the absolute scale but the depends on the period. So using a global orbit since some epoch as @ironMann was suggesting will not work unless we recalibrate the start_of_run orbit all the time. And doing this is exactly equivalent to using this formula with specific orbit duration.

Anyway, since putting the CTP TF_time or even run_start_time to every FLP might be problematic, perhaps we can do easier for the query to ccdb: it the timestamp of building the message on Stfbuilder is guaranteed to be within the current run start CTPtime and that of the next run (i.e. we are not incredibly quick in starting a new run after stopping the previous one) then we can live with 
ccdb service "get me TFtimestamp for this tfOrbit for the run to which includes this time".
Then the only requirement to CTP will be to register at the start of the run a ccdb object with abs.time of 1st orbit + 1st_orbitID + orbit_duration, this should be done anyway.

**Comment by ktf** [comment]
Ok, sounds good to me. @mkrzewic can you change the name to tfOrbit and then we merge?

**Comment by shahor02** [comment]
@ktf so, will you use the "approximate" time-stamp from the DataProcessingHeader or you want to move it to the DataHeader?

**Comment by ktf** [comment]
To start with I will simply copy the tfOrbit to the DataProcessingHeader so that we unblock (at least partially) the integration with the STFS. We can then discuss what is the best way to proceed for the rest. We could keep the DataProcessingHeader only for things which we know will need precise timestamp or subtimeframe duration.

**Comment by mkrzewic** [comment]
how about "firstTForibit" as the name? a bit more clear what is meant. Other than this - made it 32bit, and it seems everybody agrees this is the bare minimum we need. BUT: personally i'd like to add another 32bit(?) field for run number for completeness - if we agree then rather sooner than later.

**Comment by ktf** [comment]
Maybe I too anglophone but `firstTForibit` sounds like the "orbit of the first timeframe (of the run)" to me `tfFirstOrbit`? 

We can reserve another 32 bits, yes. I would actually have duration in orbits and subTimeframe orbit, as I said, but we can discuss this separately, I guess.

**Comment by mkrzewic** [comment]
"firstTForbit" lets the "TF" be capitalized, looks arguably more readable?

**Comment by ktf** [comment]
As long as US colleagues do not get confused... ;-)

**Comment by shahor02** [comment]
''andThisIsTheFirstOrbitOfTheCurrentTF''

**Comment by ktf** [comment]
`makeTimeframesGreatAgain`

**Comment by mkrzewic** [comment]
I think for it to be the orbit of the first TF it would need to be "first TF's orbit" - " first TF orbit" conveys the right message AFAIK

**Comment by ktf** [comment]
Saxon genitive is not used with objects.. Anyways... I do not care. ;-) Whatever compiles first.

**Comment by shahor02** [comment]
@ktf I see that the ``DataHeader::firstTForbit`` injected into DPL (e.g. by the ``o2-raw-file-reader``) is not propagated downstream: I would expect the output of the device to copy in its header the firstTForbit of the input, but this is not done. Is this planned? 
I am using the orbit from header to create the CTF header and CTF output file name, but currently all ``firstTForbit``are 0 except the headers of the 1st generation, so with multiple TFs the CTFs are overwritten.

**Comment by ktf** [comment]
yes, I need to take care of that and simply drop the field from the `DataProcessingHeader`. I will try to look at it later today.
